<div style="background: black; padding: 10px 250px"><img src="https://www.veldikompetens.se/wp-content/themes/consid/static/icons/VeldiKompetens_Logo_Web_Negative.svg" title="Veldi kompetens" /></div>

<hr><h1><center>Exercise 3a - Pandas Basics </center></h1>

<h3>Instructions </h3>

<p>In this exercise you will learn about the basics of Pandas; Loading a dataset, rows, columns, showing header, printing a set of lines and for example printing a row or a column. The dataset we will be using in this exercise and forwards is about different types of chocolates and how professionals have rated them. Lets get to know the dataset! 

If you want to read some more on the dataset you can find information here https://www.kaggle.com/rtatman/chocolate-bar-ratings</p>

<h3> 1. Setup </h3>

Neccesary libraries:

In [2]:
import numpy as np
import pandas as pd

<h3> 2. Pandas Basics</h3>
<h4> 2.1 Loading the dataset </h4>
<p> Note: it is a common practice to name the variable df which stands for Data Frame. Since you will be using this variable A LOT it is useful to have a concise name just as we do with numpy and pandas. Lets take a look!  </p>
<p>IMPORTANT: You need to download the dataset first and put it into the same folder as the current notebook you are working in to run the code below. Dataset can be found on itslearning </p>

In [3]:
df = pd.read_csv("flavors_of_cacao.csv", na_values = [" "])
df


Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
0                   A. Morin                       Agua Grande  1876.0   
1                   A. Morin                             Kpime  1676.0   
2                   A. Morin                            Atsane  1676.0   
3                   A. Morin                             Akata  1680.0   
4                   A. Morin                            Quilla  1704.0   
...                      ...                               ...     ...   
1794                  Zotter                              Peru   647.0   
1795                  Zotter                             Congo   749.0   
1796                  Zotter                      Kerala State   749.0   
1797                  Zotter                      Kerala State   781.0   
1798                  Zotter                Brazil, Mitzi Blue   486.0   

      Review Date Cocoa Percent Company Location  Rating  Bean Type  \
0          2016.0           63%           France    3.75        NaN   
1          2015.0           70%           France    2.75              
2          2015.0           70%           France    3.00              
3          2015.0           70%           France    3.50              
4          2015.0           70%           France    3.50              
...           ...           ...              ...     ...        ...   
1794       2011.0           70%          Austria    3.75              
1795       2011.0           65%          Austria    3.00  Forastero   
1796       2011.0           65%          Austria    3.50  Forastero   
1797       2011.0           62%          Austria    3.25              
1798       2010.0           65%          Austria    3.00              

     Broad Bean Origin  
0             Sao Tome  
1                 Togo  
2                 Togo  
3                 Togo  
4                 Peru  
...                ...  
1794              Peru  
1795             Congo  
1796             India  
1797             India  
1798            Brazil  

[1799 rows x 9 columns]

Looks nice right? Pandas visualizes the csv file and gives  you a ton of information just from reading the csv file. Truly an amazing library.

<h4>2.2 Some useful commands </h4>

In [5]:
print(f"The type is: {type(df)} \n")
print(f"The datatypes for each column: \n {df.dtypes}")

The type is: <class 'pandas.core.frame.DataFrame'> 

The datatypes for each column: 
 Company(Maker-if known)               object
 Specific Bean Origin or Bar Name     object
REF                                  float64
Review Date                          float64
Cocoa Percent                         object
Company Location                      object
Rating                               float64
Bean Type                             object
Broad Bean Origin                     object
dtype: object


<p>Before we proceed ensure that you have examined the dataset above and have a feel for what type of data value should be present in each column. Now lets move onto learning some useful commands! </p>

In [6]:
# Note; Might look really ugly but the notebook cant put everything on one line! The \ is a separator to signify same print
# But new line
print("From the bottom: ",df.tail(1), "\n")
print("From the top: ", df.head(1), "\n")

# Can also do as below to make use of Pandas awesome integration with jupyter and get a real table!
df.head(1)

From the bottom:       Company(Maker-if known)  Specific Bean Origin or Bar Name    REF  \
1798                  Zotter                Brazil, Mitzi Blue  486.0   

      Review Date Cocoa Percent Company Location  Rating Bean Type  \
1798       2010.0           65%          Austria     3.0             

     Broad Bean Origin  
1798            Brazil   

From the top:    Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
0                A. Morin                       Agua Grande  1876.0   

   Review Date Cocoa Percent Company Location  Rating Bean Type  \
0       2016.0           63%           France    3.75       NaN   

  Broad Bean Origin  
0          Sao Tome   



Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
0                A. Morin                       Agua Grande  1876.0   

   Review Date Cocoa Percent Company Location  Rating Bean Type  \
0       2016.0           63%           France    3.75       NaN   

  Broad Bean Origin  
0          Sao Tome

Something that wont be showcased here but is worth pointing out is that pandas also has a functionality called df.to_numpy() which makes ur entire dataframe into a numpy array. Its quick and cheap in terms of computational cost for inputs of the same type and is ideally used for usage on say all columns sharing type. Anyways, onwards we go!

We just saw the head (i.e the top) and the tail (i.e the bottom) of the dataframe. To access all values (indexes or rows) and columns there are also commands for that purpose as can be seen below

In [7]:
print(df.columns)
print(df.index)

Index(['Company(Maker-if known)', ' Specific Bean Origin or Bar Name', 'REF',
       'Review Date', 'Cocoa Percent', 'Company Location', 'Rating',
       'Bean Type', 'Broad Bean Origin'],
      dtype='object')
RangeIndex(start=0, stop=1799, step=1)


Next up is the most descriptive functionality of all: describe

In [8]:
# Does not include every column as only descriptive summaries of descriptive values can be showcased
print(df.describe())

               REF  Review Date      Rating
count  1796.000000  1796.000000  1796.00000
mean   1035.985523  2012.325724     3.18569
std     552.742940     2.926438     0.47804
min       5.000000  2006.000000     1.00000
25%     576.000000  2010.000000     2.75000
50%    1071.000000  2013.000000     3.25000
75%    1502.000000  2015.000000     3.50000
max    1952.000000  2017.000000     5.00000


There are several ways of sorting your data, either by index or by column values. One of the more common ones is df.sort_values(). Note we use copy here just as we do with numpy!

In [9]:
sorted_df = df.copy().sort_values(("REF"))
sorted_df

Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
882            Jacque Torres               Trinatario Treasure     5.0   
1330                  Pralus                          Trinidad     5.0   
1331                  Pralus         Madagascar, Nosy Be Isle.     5.0   
1329                  Pralus               Sao Tome & Principe     5.0   
743   Green & Black's (ICAM)                              Dark    15.0   
...                      ...                               ...     ...   
963              Letterpress      Maya Mtn, Batch 18, Heirloom  1952.0   
539              Dick Taylor                    Lanquin Estate  1952.0   
67                       NaN                               NaN     NaN   
116                      NaN                               NaN     NaN   
142                      NaN                               NaN     NaN   

      Review Date Cocoa Percent Company Location  Rating   Bean Type  \
882        2006.0           71%           U.S.A.    2.00  Trinitario   
1330       2006.0           75%           France    3.50  Trinitario   
1331       2006.0           75%           France    3.75     Criollo   
1329       2006.0           75%           France    3.50   Forastero   
743        2006.0           70%             U.K.    2.50               
...           ...           ...              ...     ...         ...   
963        2017.0           70%           U.S.A.    3.25               
539        2017.0           70%           U.S.A.    3.75               
67            NaN           NaN              NaN     NaN         NaN   
116           NaN           NaN              NaN     NaN         NaN   
142           NaN           NaN              NaN     NaN         NaN   

        Broad Bean Origin  
882                 Ghana  
1330             Trinidad  
1331           Madagascar  
1329  Sao Tome & Principe  
743                        
...                   ...  
963                Belize  
539             Guatemala  
67                    NaN  
116                   NaN  
142                   NaN  

[1799 rows x 9 columns]

<h4>2.3 Getting values </h4>

In [12]:
# Returns the values from specified column
df["REF"]

0       1876.0
1       1676.0
2       1676.0
3       1680.0
4       1704.0
         ...  
1794     647.0
1795     749.0
1796     749.0
1797     781.0
1798     486.0
Name: REF, Length: 1799, dtype: float64

In [13]:
# Each row from 0 to 1
df[0:2]

Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
0                A. Morin                       Agua Grande  1876.0   
1                A. Morin                             Kpime  1676.0   

   Review Date Cocoa Percent Company Location  Rating Bean Type  \
0       2016.0           63%           France    3.75       NaN   
1       2015.0           70%           France    2.75             

  Broad Bean Origin  
0          Sao Tome  
1              Togo

Two other common ways of accessing data is by usage of loc and iloc. In pandas loc is used to get rows and/or columns with particular label, whilst iloc gets them via integer locations or index

In [14]:
# Get a specific row by the INDEX it is at (Note when we loaded the data pandas generated indexes for each instance) 
sorted_df.iloc[3]

Company(Maker-if known)                           Pralus
 Specific Bean Origin or Bar Name    Sao Tome & Principe
REF                                                  5.0
Review Date                                       2006.0
Cocoa Percent                                        75%
Company Location                                  France
Rating                                               3.5
Bean Type                                      Forastero
Broad Bean Origin                    Sao Tome & Principe
Name: 1329, dtype: object

In [15]:
# Get a specific row from the label (this means that originally this object was given index 3 as can be seen by name)
sorted_df.loc[3]

Company(Maker-if known)              A. Morin
 Specific Bean Origin or Bar Name       Akata
REF                                    1680.0
Review Date                            2015.0
Cocoa Percent                             70%
Company Location                       France
Rating                                    3.5
Bean Type                                    
Broad Bean Origin                        Togo
Name: 3, dtype: object

<h4>2.4 Time for some Q/A! </h4> 

In [17]:
#Q1: Output the first three rows in sorted_df and df respectively
# here output first three of df
df[0:3]

Company(Maker-if known)  Specific Bean Origin or Bar Name     REF  \
0                A. Morin                       Agua Grande  1876.0   
1                A. Morin                             Kpime  1676.0   
2                A. Morin                            Atsane  1676.0   

   Review Date Cocoa Percent Company Location  Rating Bean Type  \
0       2016.0           63%           France    3.75       NaN   
1       2015.0           70%           France    2.75             
2       2015.0           70%           France    3.00             

  Broad Bean Origin  
0          Sao Tome  
1              Togo  
2              Togo

Expected output: 
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Company&nbsp; (Maker-if known)</th>
      <th>Specific Bean Origin or Bar Name</th>
      <th>REF</th>
      <th>Review Date</th>
      <th>Cocoa Percent</th>
      <th>Company Location</th>
      <th>Rating</th>
      <th>Bean Type</th>
      <th>Broad Bean Origin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A. Morin</td>
      <td>Agua Grande</td>
      <td>1876</td>
      <td>2016</td>
      <td>63%</td>
      <td>France</td>
      <td>3.75</td>
      <td></td>
      <td>Sao Tome</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A. Morin</td>
      <td>Kpime</td>
      <td>1676</td>
      <td>2015</td>
      <td>70%</td>
      <td>France</td>
      <td>2.75</td>
      <td></td>
      <td>Togo</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A. Morin</td>
      <td>Atsane</td>
      <td>1676</td>
      <td>2015</td>
      <td>70%</td>
      <td>France</td>
      <td>3.00</td>
      <td></td>
      <td>Togo</td>
    </tr>
  </tbody>
</table>

In [18]:
# here output first three of sorted_df
sorted_df[0:3]


Company(Maker-if known)  Specific Bean Origin or Bar Name  REF  \
882            Jacque Torres               Trinatario Treasure  5.0   
1330                  Pralus                          Trinidad  5.0   
1331                  Pralus         Madagascar, Nosy Be Isle.  5.0   

      Review Date Cocoa Percent Company Location  Rating   Bean Type  \
882        2006.0           71%           U.S.A.    2.00  Trinitario   
1330       2006.0           75%           France    3.50  Trinitario   
1331       2006.0           75%           France    3.75     Criollo   

     Broad Bean Origin  
882              Ghana  
1330          Trinidad  
1331        Madagascar

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Company(Maker-if known)</th>
      <th>Specific Bean Origin or Bar Name</th>
      <th>REF</th>
      <th>Review Date</th>
      <th>Cocoa Percent</th>
      <th>Company Location</th>
      <th>Rating</th>
      <th>Bean Type</th>
      <th>Broad Bean Origin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>882</th>
      <td>Jacque Torres</td>
      <td>Trinatario Treasure</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>71%</td>
      <td>U.S.A.</td>
      <td>2.00</td>
      <td>Trinitario</td>
      <td>Ghana</td>
    </tr>
    <tr>
      <th>1330</th>
      <td>Pralus</td>
      <td>Trinidad</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>75%</td>
      <td>France</td>
      <td>3.50</td>
      <td>Trinitario</td>
      <td>Trinidad</td>
    </tr>
    <tr>
      <th>1331</th>
      <td>Pralus</td>
      <td>Madagascar, Nosy Be Isle.</td>
      <td>5.0</td>
      <td>2006.0</td>
      <td>75%</td>
      <td>France</td>
      <td>3.75</td>
      <td>Criollo</td>
      <td>Madagascar</td>
    </tr>
  </tbody>
</table>

In [19]:
#Q2: Get the Broad Bean origin for the third item from the top from the sorted_df (i.e Ghana)
# Hint: Use iloc to get a specific row from the top, syntax should be sorted_df.iloc[smthing][column name]
sorted_df[0:3]["Broad Bean Origin"]

882          Ghana
1330      Trinidad
1331    Madagascar
Name: Broad Bean Origin, dtype: object

Great job! We are now done with the basic introduction to Pandas! Lets proceed to the next exercise where we will learn to work some more with the dataset and reason a bit more on it